src/main/java/com/example/Library/services/BookService.java

```diff
@@ -69,4 +73,26 @@ public Optional<Book> returnBook(long id){
                     return bookRepository.save(book);
                 });
     }
+
+    public Book filter(String idOrTitle) {
+        char[] chars = idOrTitle.toCharArray();
+        for (char symbol : chars) {
+            int x = symbol;
+            if ((x >= 97 && x <= 122) || (x >= 65 && x <= 90) || (x >= 1040 && x <= 1071) || (x >= 1073 && x <= 1103)) {
+                return bookRepository.getBookByTitle(idOrTitle)
+                        .orElseThrow(() -> new BookNotFoundByTitleException(idOrTitle));
+            }
+        }
+        long id = Long.parseLong(idOrTitle);
+        return bookRepository.findById(id)
+                .orElseThrow(() -> new BookNotFoundByIdException(id));
+    }
+
+    public Iterable<Book> getAllBooksWithFilterByTitle(String title) {
+        if (title != null) {
```

Comment:
Minor: this if statement could be moved to the controller, there it might be a bit more obvious how query params influence the result.